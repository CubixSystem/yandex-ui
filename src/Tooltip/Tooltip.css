.Tooltip {
    z-index: 1500;

    font-family: var(--tooltip-font-family);
}

.Tooltip.Popup2::before {
    content: none;
}
