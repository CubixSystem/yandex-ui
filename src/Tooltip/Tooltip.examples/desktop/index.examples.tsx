import { withKnobs } from '@storybook/addon-knobs';

import { EXAMPLE_DESKTOP_TOKEN, createDecorators, parameters } from '../examples-config';

export * from './playground.examples';
export * from './direction.examples';
export * from './size.examples';
export * from './state.examples';
export * from './stateful.examples';

export default {
    title: EXAMPLE_DESKTOP_TOKEN,
    decorators: [withKnobs, ...createDecorators({ scope: 'desktop' })],
    parameters,
};
