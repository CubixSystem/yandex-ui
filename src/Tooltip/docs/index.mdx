---
id: Tooltip
path: /components/tooltip

title: Tooltip
description: Компонент, использующийся для создания всплывающих подсказок.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Tooltip
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Tooltip.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/L-E-G-O?node-id=0%3A1

tabs:
 - Examples
 - Usage
 - Props
 - Accessibility
---
