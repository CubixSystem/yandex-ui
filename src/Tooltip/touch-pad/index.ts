export * from '../Tooltip@touch-pad';
export * from '../_size/Tooltip_size_l';
export * from '../_size/Tooltip_size_m';
export * from '../_size/Tooltip_size_s';
export * from '../_view/Tooltip_view_default';
