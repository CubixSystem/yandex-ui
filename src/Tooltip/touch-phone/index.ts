export * from '../Tooltip@touch-phone';
export * from '../_size/Tooltip_size_l';
export * from '../_size/Tooltip_size_m';
export * from '../_size/Tooltip_size_s';
export * from '../_view/Tooltip_view_default';
