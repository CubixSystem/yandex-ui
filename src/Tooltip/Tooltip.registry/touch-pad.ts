import { Registry } from '@bem-react/di';
import { compose } from '@bem-react/core';

import { Popup as PopupTouchPad } from '../../Popup/Popup@touch-pad';
import { withViewDefault } from '../../Popup/_view/Popup_view_default';
import { withTargetAnchor } from '../../Popup/_target/Popup_target_anchor';
import { cnTooltip } from '../Tooltip';

const Popup = compose(
    withViewDefault,
    withTargetAnchor,
)(PopupTouchPad);

export const registry = new Registry({ id: cnTooltip() });

registry.set('Popup', Popup);
