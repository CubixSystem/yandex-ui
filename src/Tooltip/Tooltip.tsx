import React, { FC } from 'react';
import { cn } from '@bem-react/classname';
import { useComponentRegistry } from '@bem-react/di';

import { IPopupProps, Direction } from '../Popup/Popup';
import { IPopupTargetAnchorProps } from '../Popup/_target/Popup_target_anchor';
import { TooltipBackdrop as Backdrop } from './Backdrop/Tooltip-Backdrop';
import { TooltipRegistry } from './Tooltip.registry/interface';
import './Tooltip.css';

export { Direction } from '../Popup/Popup';

export const cnTooltip = cn('Tooltip');

type PartialPopupProps = Pick<
    IPopupProps,
    'keepMounted' | 'className' | 'hasTail' | 'innerRef' | 'zIndex' | 'visible' | 'scope' | 'onClose' | 'onClick'
>;

type PartialPopupTargetAnchorProps = Pick<
    IPopupTargetAnchorProps,
    'mainOffset' | 'secondaryOffset' | 'tailOffset' | 'anchor'
>;

export type TooltipProps = PartialPopupProps &
    PartialPopupTargetAnchorProps & {
        /**
         * Визуальное состояние подсказки
         */
        state?: 'warning' | 'alert' | 'success';

        /**
         * Направление для раскрытия подсказки
         */
        direction?: Direction | Direction[];

        /**
         * Уникальный id подсказки
         */
        id?: string;
    };

/**
 * Компонент используется для создания всплывающих подсказок.
 *
 * @param {TooltipProps} props Свойства компонента.
 */
export const Tooltip: FC<TooltipProps> = ({
    anchor,
    children,
    className,
    direction,
    hasTail,
    id,
    innerRef,
    keepMounted,
    mainOffset,
    scope,
    secondaryOffset,
    state,
    tailOffset,
    visible,
    zIndex,
    onClose,
    onClick,
}) => {
    const defaultMainOffset = hasTail ? 0 : 4;
    const { Popup } = useComponentRegistry<TooltipRegistry>(cnTooltip());
    const directions = Array.isArray(direction) ? direction : [direction as Direction];

    return (
        <Popup
            onClose={onClose}
            onClick={onClick}
            anchor={anchor}
            className={cnTooltip({ visible, state }, [className])}
            directions={direction ? directions : undefined}
            hasTail={hasTail}
            innerRef={innerRef}
            keepMounted={keepMounted}
            mainOffset={mainOffset || defaultMainOffset}
            scope={scope}
            secondaryOffset={secondaryOffset}
            tailOffset={tailOffset}
            target="anchor"
            view="default"
            visible={visible}
            zIndex={zIndex}
            // prettier-ignore
            unstable_onRenderTail={(tail) => (
                <Backdrop>{hasTail && tail}</Backdrop>
            )}
        >
            <div role="tooltip" id={id} className={cnTooltip('Content')}>
                {children}
            </div>
        </Popup>
    );
};

Tooltip.displayName = 'Tooltip';
