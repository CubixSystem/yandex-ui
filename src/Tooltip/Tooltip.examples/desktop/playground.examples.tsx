import React, { useState, useRef } from 'react';
import { select, boolean, number } from '@storybook/addon-knobs';

import { cnTheme } from '../../../Theme';
import { presets, presetsKeys } from '../../../Theme/presets';
import { Tooltip, Direction } from '../../Tooltip.bundle/desktop';
import { Button } from '../../../Button/Button.bundle/desktop';

const DIRECTIONS: Direction[] = [
    'top-left',
    'top-center',
    'top-right',
    'right-top',
    'right-center',
    'right-bottom',
    'bottom-right',
    'bottom-center',
    'bottom-left',
    'left-bottom',
    'left-center',
    'left-top',
];

export const Playground = () => {
    const scopeRef = useRef<HTMLDivElement>(null);
    const buttonRef = useRef<HTMLButtonElement>(null);
    const [visible, setVisible] = useState(false);

    const preset = select('theme-preset', presetsKeys, 'default');

    const size = select('size', ['s', 'm', 'l'], 'm') as any;
    const state = select('state', ['', 'warning', 'alert', 'success'], '') as any;
    const direction = select('direction', DIRECTIONS, 'right-center') as any;
    const hasTail = boolean('hasTail', true);
    const mainOffset = number('mainOffset', 0);
    const secondaryOffset = number('secondaryOffset', 0);
    const tailOffset = number('tailOffset', 0);

    return (
        <div
            ref={scopeRef}
            className={cnTheme(presets[preset])}
            style={{ position: 'relative', backgroundColor: 'var(--color-bg-default)' }}
        >
            <Button innerRef={buttonRef} view="default" size="m" onClick={() => setVisible(!visible)}>
                Target
            </Button>
            <Tooltip
                key={direction}
                hasTail={hasTail}
                mainOffset={mainOffset}
                secondaryOffset={secondaryOffset}
                tailOffset={tailOffset}
                direction={direction}
                anchor={buttonRef}
                scope={scopeRef}
                visible={visible}
                view="default"
                size={size}
                state={state}
            >
                Добавить в избранное
            </Tooltip>
        </div>
    );
};
