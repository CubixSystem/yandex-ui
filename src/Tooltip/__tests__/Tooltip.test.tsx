import React, { createRef, FC, RefObject } from 'react';

import { createClientRender, screen, fireEvent, act } from '../../internal/testing';
import { Tooltip as TooltipCommon } from '../Tooltip';
import { Tooltip as TooltipDesktop, TooltipStateful as TooltipStatefulDesktop } from '../desktop';
import { Tooltip as TooltipTouchPad, TooltipStateful as TooltipStatefulTouchPad } from '../touch-phone';
import { Tooltip as TooltipTouchPhone, TooltipStateful as TooltipStatefulTouchPhone } from '../touch-pad';

const platforms = [
    ['desktop', TooltipDesktop, TooltipStatefulDesktop],
    ['touch-pad', TooltipTouchPad, TooltipStatefulTouchPad],
    ['touch-phone', TooltipTouchPhone, TooltipStatefulTouchPhone],
] as const;

describe.each(platforms)('Tooltip@%s', (_platform, Tooltip, TooltipStatefull) => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        render(
            <Tooltip visible anchor={createRef()}>
                Добавить в избранное
            </Tooltip>,
        );

        expect(document.body).toMatchSnapshot();
    });

    test('должен быть установлен корректный displayName', () => {
        expect(TooltipCommon.displayName).toBe('Tooltip');
    });

    test('должен устанавливать ссылку на корневой DOM элемент', async () => {
        const innerRef = createRef<HTMLDivElement>();
        render(<Tooltip visible innerRef={innerRef} anchor={createRef()} />);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен устанавливать модификатор visible если открыт при инициализации', () => {
        const innerRef = createRef<HTMLDivElement>();
        render(<Tooltip innerRef={innerRef} visible anchor={createRef()} />);

        expect(innerRef.current).toHaveClass('Tooltip_visible');
    });

    test('должен устанавливать/удалять модификатор visible при открытии/закрытии', () => {
        const innerRef = createRef<HTMLDivElement>();
        const { setProps } = render(<Tooltip innerRef={innerRef} anchor={createRef()} />);

        setProps({ visible: true });
        expect(innerRef.current).toHaveClass('Tooltip_visible');

        setProps({ visible: false });
        expect(innerRef.current).not.toHaveClass('Tooltip_visible');
    });

    describe('stateful', () => {
        beforeEach(() => {
            jest.useFakeTimers();
        });

        afterEach(() => {
            jest.useRealTimers();
        });

        it('должен установить innerRef', () => {
            const Anchor: FC<{ innerRef: RefObject<HTMLButtonElement> }> = ({ innerRef }) => {
                return (
                    <button ref={innerRef} data-testid="trigger">
                        trigger
                    </button>
                );
            };
            const ref = createRef<HTMLButtonElement>();

            render(
                <TooltipStatefull defaultVisible content={<span data-testid="tooltip">content</span>}>
                    <Anchor innerRef={ref}>trigger</Anchor>
                </TooltipStatefull>,
            );

            expect(ref.current).not.toBe(null);
        });

        it('должен открыть tooltip с задержкой', () => {
            render(
                <TooltipStatefull
                    content={<span data-testid="tooltip">content</span>}
                    trigger="click"
                    openDelay={1000}
                    keepMounted={false}
                >
                    <span data-testid="trigger">trigger</span>
                </TooltipStatefull>,
            );

            fireEvent.click(screen.getByTestId('trigger'));
            expect(screen.queryByTestId('tooltip')).not.toBeInTheDocument();

            act(() => {
                jest.runOnlyPendingTimers();
            });

            expect(screen.getByTestId('tooltip')).toBeInTheDocument();
        });

        it('должен закрыть tooltip с задержкой', () => {
            render(
                <TooltipStatefull
                    defaultVisible
                    content={<span data-testid="tooltip">content</span>}
                    trigger="click"
                    closeDelay={1000}
                    keepMounted={false}
                >
                    <span data-testid="trigger">trigger</span>
                </TooltipStatefull>,
            );

            fireEvent.click(screen.getByTestId('trigger'));
            expect(screen.getByTestId('tooltip')).toBeInTheDocument();

            act(() => {
                jest.runOnlyPendingTimers();
            });

            expect(screen.queryByTestId('tooltip')).not.toBeInTheDocument();
        });
    });
});
