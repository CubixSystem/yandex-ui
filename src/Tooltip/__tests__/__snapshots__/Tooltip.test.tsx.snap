// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<body>
  <div />
  <div
    class="Popup2 Popup2_visible Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
    style="top: -9999px; left: -9999px; position: fixed;"
  >
    <div
      class="Tooltip-Content"
      role="tooltip"
    >
      Добавить в избранное
    </div>
    <div
      class="Tooltip-Backdrop"
    />
  </div>
</body>
`;

exports[`Tooltip@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<body>
  <div />
  <div
    class="Popup2 Popup2_visible Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
    style="top: -9999px; left: -9999px; position: fixed;"
  >
    <div
      class="Tooltip-Content"
      role="tooltip"
    >
      Добавить в избранное
    </div>
    <div
      class="Tooltip-Backdrop"
    />
  </div>
</body>
`;

exports[`Tooltip@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<body>
  <div />
  <div
    class="Popup2 Popup2_visible Popup2_target_anchor Popup2_view_default Tooltip Tooltip_visible"
    style="top: -9999px; left: -9999px; position: fixed;"
  >
    <div
      class="Tooltip-Content"
      role="tooltip"
    >
      Добавить в избранное
    </div>
    <div
      class="Tooltip-Backdrop"
    />
  </div>
</body>
`;
